tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -212,7 +218,383 @@
+      return Collections.EMPTY_LIST;
+    }
+
+    Operand<T> lLabels = labels;
+    Operand<T> lPredictions = predictions;
+    Operand<T> lSampleWeight = sampleWeight;
+
+    Operand<TInt32> numThresholds;
+    Operand<TBool> oneThresh;
+    if (multiLabel) {
+      numThresholds = tf.shape.size(lLabels, tf.constant(0));
+      oneThresh = tf.math.equal(tf.constant(1), tf.constant(thresholds.length));
+    } else {
+      // TODO handle Ragged Tensors????
+      // [y_pred,
+      //    y_true], _ = ragged_assert_compatible_and_get_flat_values([y_pred, y_true],
+      //                                                   sampleWeights)
+      numThresholds = tf.constant(thresholds.length);
+      oneThresh = tf.constant(true);
+    }
+
+    List<Op> controlOps = new ArrayList<>();
+    Operand<TInt32> axes = allAxes(tf, lPredictions);
+    controlOps.add(
+        tf.withSubScope("updateConfusionMatrixVariables-1")
+            .assertThat(
+                tf.reduceAll(
+                    tf.math.greaterEqual(
+                        lPredictions, cast(tf, tf.constant(0), lPredictions.type())),
+                    axes),
+                Collections.singletonList(tf.constant("predictions must be >= 0"))));
+    controlOps.add(
+        tf.withSubScope("updateConfusionMatrixVariables-2")
+            .assertThat(
+                tf.reduceAll(
+                    tf.math.lessEqual(lPredictions, cast(tf, tf.constant(1), lPredictions.type())),
+                    axes),
+                Collections.singletonList(tf.constant("predictions must be <= 1"))));
+
+    LossTuple<T> result =
+        LossesHelper.squeezeOrExpandDimensions(tf, lLabels, lPredictions, lSampleWeight);
+    lPredictions = result.getTarget();
+    lLabels = result.getLabels();
+    lSampleWeight = result.getSampleWeights();
+
+    if (!lPredictions.shape().isCompatibleWith(lLabels.shape()))
+      throw new IllegalArgumentException(
+          String.format(
+              "Shapes %s and %s are incompatible)",
+              lPredictions.shape().toString(), lLabels.asOutput().shape().toString()));
+
+    if (topK != null) {
+      lPredictions = filterTopK(tf, lPredictions, topK);
+    }
+
+    if (classId != null) {
+      lLabels = tf.squeeze(tf.gather(lLabels, tf.constant(new int[] {classId}), tf.constant(1)));
+      lPredictions =
+          tf.squeeze(tf.gather(lPredictions, tf.constant(new int[] {classId}), tf.constant(1)));
+      lLabels = tf.expandDims(lLabels, tf.constant(0));
+      lPredictions = tf.expandDims(lPredictions, tf.constant(0));
+    }
+    org.tensorflow.op.core.Shape<TInt32> predShape = tf.shape(lPredictions);
+    Operand<TInt32> numPredictions =
+        tf.reshape(tf.shape.size(lPredictions, tf.constant(0)), tf.constant(Shape.scalar()));
+    Operand<TInt32> numLabels =
+        tf.select(
+            tf.math.equal(tf.shape.numDimensions(predShape), tf.constant(1)),
+            tf.constant(1),
+            tf.reduceProd(
```

Comment:
This code does extra work to accommodate prediction/label rank > 2, but it's not clear to me that our earlier code supports such a situation:
* In the multi-label case, we've already asserted that our labels are rank 2.
* Our handling of `classId` seems semantically odd if our predictions and labels have rank > 2. In that case, `classId`   limits consideration to a slice of predictions/labels for each example rather than (as its name suggests) a single prediction/label for each example.

Reply:
The only restriction is on `multiLabel`, rank must be 2.

Also, why couldn't class features be represented with multi-dimensional input? Class (classID) is just a vertical slice of the feature set, whether it be  a 1D or an nD slice.

Reply:
👍 done